File: backend/app/schemas/graph.py
import uuid

from pydantic import BaseModel


class GraphNodeResponse(BaseModel):
    id: str
    type: str
    label: str
    metadata: dict


class GraphEdgeResponse(BaseModel):
    id: str
    from_type: str
    from_id: str
    to_type: str
    to_id: str
    relation: str
    source: str
    reason: str
    metadata: dict


class GraphResponse(BaseModel):
    nodes: list[GraphNodeResponse]
    edges: list[GraphEdgeResponse]


class GraphRebuildRequest(BaseModel):
    project_id: uuid.UUID | None = None
    asset_id: uuid.UUID | None = None


class GraphRebuildTriggerResponse(BaseModel):
    status: str


# Phase 24 -- Attack Path Analysis. No exploitability/probability score
# anywhere in this shape, deliberately: `disclaimer` is the one thing every
# caller (API consumer or frontend) is guaranteed to see alongside the
# paths themselves.
class AttackPath(BaseModel):
    hops: int
    nodes: list[GraphNodeResponse]
    edges: list[GraphEdgeResponse]


class AttackPathsResponse(BaseModel):
    disclaimer: str
    seed: GraphNodeResponse
    truncated: bool
    paths: list[AttackPath]

File: frontend/app/types/graph.ts
// Mirrors backend/app/schemas/graph.py exactly. Moved out of
// SecurityGraph.vue so the same shapes are available to future consumers
// (e.g. a global /graph search page in 18c) without re-declaring them.
export type GraphNodeType = "ASSET" | "FINDING" | "CVE" | "SERVICE" | "TECHNOLOGY";

export interface GraphNode {
  id: string;
  type: GraphNodeType;
  label: string;
  metadata: Record<string, unknown>;
}

export interface GraphEdge {
  id: string;
  from_type: string;
  from_id: string;
  to_type: string;
  to_id: string;
  relation: string;
  source: string;
  reason: string;
  metadata: Record<string, unknown>;
}

export interface GraphResponse {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

// Phase 24 -- Attack Path Analysis. Mirrors backend/app/schemas/graph.py's
// AttackPath/AttackPathsResponse exactly. `disclaimer` is always present
// on the response, deliberately -- never trimmed on the assumption the UI
// already shows one of its own.
export interface AttackPath {
  hops: number;
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export interface AttackPathsResponse {
  disclaimer: string;
  seed: GraphNode;
  truncated: boolean;
  paths: AttackPath[];
}
